src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -0,0 +1,107 @@
+package org.sagebionetworks.bridge.services;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.sagebionetworks.bridge.validators.PermissionValidator.INSTANCE;
+
+import org.sagebionetworks.bridge.BridgeUtils;
+import org.sagebionetworks.bridge.dao.PermissionDao;
+import org.sagebionetworks.bridge.exceptions.EntityNotFoundException;
+import org.sagebionetworks.bridge.models.accounts.Account;
+import org.sagebionetworks.bridge.models.accounts.AccountId;
+import org.sagebionetworks.bridge.models.accounts.AccountRef;
+import org.sagebionetworks.bridge.models.permissions.EntityType;
+import org.sagebionetworks.bridge.models.permissions.Permission;
+import org.sagebionetworks.bridge.models.permissions.PermissionDetail;
+import org.sagebionetworks.bridge.validators.Validate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@Component
+public class PermissionService {
+    
+    private PermissionDao permissionDao;
+    
+    private AccountService accountService;
+    
+    @Autowired
+    final void setPermissionDao(PermissionDao permissionDao) {
+        this.permissionDao = permissionDao;
+    }
+    
+    @Autowired
+    final void setAccountService(AccountService accountService) {
+        this.accountService = accountService;
+    }
+    
+    // For mock testing
+    protected String generateGuid() {
+        return BridgeUtils.generateGuid();
+    }
+    
+    public PermissionDetail createPermission(String appId, Permission permission) {
+        checkArgument(isNotBlank(appId));
+        checkNotNull(permission);
+        
+        permission.setGuid(generateGuid());
+        
+        Validate.entityThrowingException(INSTANCE, permission);
+        
+        Permission createdPermission = permissionDao.createPermission(appId, permission);
+        
+        return getPermissionDetail(appId, createdPermission);
+    }
+    
+    public PermissionDetail updatePermission(String appId, Permission permission) {
+        checkArgument(isNotBlank(appId));
+        checkNotNull(permission);
+    
+        Validate.entityThrowingException(INSTANCE, permission);
+        
+        Permission updatedPermission = permissionDao.updatePermission(appId, permission);
```

Comment:
Get the permission by Guid. If it doesn't exist, throw EntityNotFound.

Overwrite existing.accessLevel with permission.accessLevel to prevent them from changing the user ID or entity. (Alternatively, check if these are changed and throw a BadRequest if they are.)

Reply:
Updated to throw EntityNotFound if the guid does not exist and overwrite all fields except for accessLevel with previous values.

Reply:
I don't see a @Version field for this entity which we normally provide. It will give you optimistic locking behavior: if someone updates a permission and then you update it, you'll get an optimistic lock exception and you can examine the new object state before deciding to proceed with your edit. In any event, version would also need to be copied over.

Reply:
We also have started adding createdOn/modifiedOn timestamps to database objects. It's not a great form of auditing but it's better than nothing and would be consistent.